Implement key resharing.

robin-near opened the pull request:
This PR implement key resharing. It is expected to work out of the box as long as we manually kick the contract into a Resharing state (by sending join and vote_join; leaving is fine too). Some of the code was adapted from #87 .

Rust-side integration tests are included to test resharing behavior including signature buffering during resharing. **Python integration test as well as real cluster test are TODO**.

Along with that are a few more related or necessary changes:
* Make the coordinator always check if the current node is not a participant, and halt in this case (until contract state change) instead of failing (and then looping).
* Make the network layer's `wait_for_ready` wait for that many connections to be *bidirectionally* connected, not just outgoing. This solves a subtle problem where nodes may repeatedly fail to carry out resharding and run into a subtle retry loop. It is also good to have - can possibly avoid hiding network issues where one side cannot accept incoming connections. Further research may be needed to improve the reliability here (#151 ).
  * Related, also make keygen and resharing loop until receiving the expected keygen/resharing task from the network; this can probably avoid unnecessary failures where other nodes have not transitioned into resharing state yet.
* Some test code refactoring to better support tests where the participant set changes.
* Add `PortSeed` to better avoid port conflicts in tests.

bowenwang1996 commented on node/src/p2p.rs:
```diff
@@ -582,6 +600,70 @@ fn verify_peer_identity(
 
+/// Struct to track bidirectional connectivity between two nodes.
+#[derive(Clone)]
+struct ConnectivityIndicator {
```
Could you add some more comments to explain how this struct works?

bowenwang1996 reviewed (approved):
Great work!

robin-near commented on node/src/coordinator.rs:
```diff
@@ -385,18 +400,109 @@ impl Coordinator {
+            }
+        };
+
+        // Delete all presignatures from the previous epoch; they are no longer usable
+        // once we reshare keys.
```
I think you said triples don't need to be deleted right?

The triples are not tied to keys, only participant IDs, and the triples carry the list of participant IDs with them, so we can still reuse them.

Well... I guess some triples would be useless so we should clear them out. Yeah. let me send a separate PR for that. Less of a urgent thing because it's only relevant for leaving.

robin-near commented on node/src/coordinator.rs:
```diff
@@ -385,18 +400,109 @@ impl Coordinator {
+            &config_file.my_near_account_id,
+        ) else {
+            tracing::info!("We are not a participant in the new epoch; doing nothing until contract state change");
+            futures::future::pending::<()>().await;
+            unreachable!()
```
Yeah I'm curious too. I'm relying on just dropping futures a lot. There's an alternative approach in general of using synchronization primitives to gracefully exit (so we don't drop futures), but that is a lot harder to get right and the code becomes sprinkled with `select!`s everywhere.

Hmm, perhaps for this future, we can just return instead of hanging. It would return a different flag saying that it should hang instead of proceeding to the next iteration. Yeah, that would probably work - I'll give it a try.

Wait, do I sound like DeepThink R1? :joy:

saketh-are commented on node/src/db.rs:
```diff
@@ -134,6 +159,17 @@ impl SecretDBUpdate {
+        if let Some(range) = range {
+            self.batch
+                .delete_range_cf(self.db.cf_handle(col), range.start(), range.end());
+            // delete_range_cf deletes ["begin_key", "end_key"), so need one more delete
```
It's surprising there isn't a nicer way of doing this. One recommendation I saw is to pass (smallest possible key, largest possible key) but that is not that great either.

robin-near replied:
Same here, copied from nearcore. I'm surprised too....

robin-near replied:
The problem is the smallest key is b"", but there does not exist a largest possible key.

robin-near commented on node/src/db.rs:
```diff
@@ -115,6 +116,30 @@ impl SecretDB {
+                .db
+                .iterator_cf(self.cf_handle(col), IteratorMode::Start);
+            let start = iter.next().transpose()?;
+            iter.set_mode(IteratorMode::End);
```
Oh, this code is copied from nearcore.

robin-near commented:
Lol I think when I deal with the futures thing, I'm uncovering a whole rabbit whole of issues. So this will take a little longer.

robin-near commented:
ok so there's some subtlety with regards to what to do when a node is not actively participating in the MPC network. If the node is not a participant, it should not be buffering the signatures in its channel. Otherwise, this would blow up the memory and when the node joins and becomes a participant, it would be retroactively trying to process all the signature requests and that is not desirable. Added some code to properly deal with that in each scenario along with addressing the `futures::future::pending` hack.

Also, earlier the test that I was doing to assert that I got a signature response was not sufficient; now I check that the response corresponds to the expected signature request. What happens is that because computation can happen before network is fully connected, when a node receives a signature request and it is the secondary leader, it might start a computation for it because it is not yet connected to the primary leader. So in tests we can often see two responses for the same signature request. I've changed it to properly wait for the correct signature response, and that fixes a flake for the faulty cluster test as well.
